File: backend-java/src/main/java/com/bffgin/backend/domain/TaskError.java
package com.bffgin.backend.domain;

/**
 * backend-rustのRestError(error.rs)と同じ分類を1つのクラスにまとめたもの。
 * JSON形状・HTTPステータス・gRPCステータスへの変換はトランスポート層(rest/grpc)がそれぞれ担う
 */
public final class TaskError extends Exception {

    public enum Kind {
        UNAUTHORIZED,
        USER_NOT_PROVISIONED,
        INVALID_REQUEST,
        INVALID_ID,
        INVALID_STATUS,
        INVALID_FINISHED_ON,
        VALIDATION,
        NOT_FOUND,
        INTERNAL,
        USER_ID_REQUIRED,
        INVALID_USER_ID
    }

    private final Kind kind;

    private TaskError(Kind kind, String message) {
        super(message);
        this.kind = kind;
    }

    public Kind kind() {
        return kind;
    }

    public static TaskError unauthorized() {
        return new TaskError(Kind.UNAUTHORIZED, "unauthorized");
    }

    public static TaskError userNotProvisioned() {
        return new TaskError(Kind.USER_NOT_PROVISIONED, "user_not_provisioned");
    }

    public static TaskError invalidRequest() {
        return new TaskError(Kind.INVALID_REQUEST, "invalid_request");
    }

    public static TaskError invalidId() {
        return new TaskError(Kind.INVALID_ID, "invalid_id");
    }

    public static TaskError invalidStatus() {
        return new TaskError(Kind.INVALID_STATUS, "invalid_status");
    }

    public static TaskError invalidFinishedOn() {
        return new TaskError(Kind.INVALID_FINISHED_ON, "invalid_finished_on");
    }

    public static TaskError validation(String message) {
        return new TaskError(Kind.VALIDATION, message);
    }

    public static TaskError notFound() {
        return new TaskError(Kind.NOT_FOUND, "not_found");
    }

    public static TaskError internal(String message) {
        return new TaskError(Kind.INTERNAL, message);
    }

    /** 外部公開API(CONTRACT.mdセクション11): user_idクエリパラメータが無い/空 */
    public static TaskError userIdRequired() {
        return new TaskError(Kind.USER_ID_REQUIRED, "user_id_required");
    }

    /** 外部公開API: user_idクエリパラメータが数値でない */
    public static TaskError invalidUserId() {
        return new TaskError(Kind.INVALID_USER_ID, "invalid_user_id");
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/domain/TaskError.kt
package com.bffgin.backend.domain

/**
 * backend-java/backend-rustのTaskError/RestErrorと同じ分類を1つのクラスにまとめたもの。
 * JSON形状・HTTPステータス・gRPCステータスへの変換はトランスポート層(rest/grpc)がそれぞれ担う。
 * Kotlinには検査例外(checked exception)が無いため、backend-javaのように呼び出し元に
 * `throws TaskError`の宣言を強制することはできないが、suspend fun群は全てこの例外だけを
 * 投げる設計を維持している
 */
class TaskError private constructor(val kind: Kind, message: String) : Exception(message) {

    enum class Kind {
        UNAUTHORIZED,
        USER_NOT_PROVISIONED,
        INVALID_REQUEST,
        INVALID_ID,
        INVALID_STATUS,
        INVALID_FINISHED_ON,
        VALIDATION,
        NOT_FOUND,
        INTERNAL,
    }

    companion object {
        fun unauthorized() = TaskError(Kind.UNAUTHORIZED, "unauthorized")
        fun userNotProvisioned() = TaskError(Kind.USER_NOT_PROVISIONED, "user_not_provisioned")
        fun invalidRequest() = TaskError(Kind.INVALID_REQUEST, "invalid_request")
        fun invalidId() = TaskError(Kind.INVALID_ID, "invalid_id")
        fun invalidStatus() = TaskError(Kind.INVALID_STATUS, "invalid_status")
        fun invalidFinishedOn() = TaskError(Kind.INVALID_FINISHED_ON, "invalid_finished_on")
        fun validation(message: String) = TaskError(Kind.VALIDATION, message)
        fun notFound() = TaskError(Kind.NOT_FOUND, "not_found")
        fun internal(message: String) = TaskError(Kind.INTERNAL, message)
    }
}
